homeassistant/components/light/template.py

```diff
@@ -0,0 +1,242 @@
+
+        if level_template is not None:
+            temp_ids = level_template.extract_entities()
+            if str(temp_ids) != '*':
+                template_entity_ids = list(set(template_entity_ids) |
+                                           set(temp_ids))
+
+        entity_ids = device_config.get(CONF_ENTITY_ID, template_entity_ids)
+
+        lights.append(
+            LightTemplate(
+                hass, device, friendly_name, state_template,
+                on_action, off_action, level_action, level_template,
+                entity_ids)
+        )
+
+    if not lights:
+        _LOGGER.error("No lights added")
+        return False
+
+    async_add_devices(lights, True)
+    return True
+
+
+class LightTemplate(Light):
+    """Representation of a templated Light, including dimmable."""
+
+    def __init__(self, hass, device_id, friendly_name, state_template,
+                 on_action, off_action, level_action, level_template,
+                 entity_ids):
+        """Initialize the light."""
+        self.hass = hass
+        self.entity_id = async_generate_entity_id(
+            ENTITY_ID_FORMAT, device_id, hass=hass)
+        self._name = friendly_name
+        self._template = state_template
+        self._on_script = Script(hass, on_action)
+        self._off_script = Script(hass, off_action)
+        self._level_script = Script(hass, level_action)
+        self._level_template = level_template
+
+        self._state = False
+        self._brightness = None
+        self._entities = entity_ids
+        self._entities.append(self.entity_id)
+
+        if self._template is not None:
+            self._template.hass = self.hass
+        if self._level_template is not None:
+            self._level_template.hass = self.hass
+
+    @property
+    def brightness(self):
+        """Return the brightness of the light."""
+        return self._brightness
+
+    @property
+    def supported_features(self):
+        """Flag supported features."""
+        if self._level_script is not None:
+            return SUPPORT_BRIGHTNESS
+
+        return 0
+
+    @property
+    def is_on(self):
+        """Return true if device is on."""
+        return self._state
+
+    @property
+    def should_poll(self):
+        """Return the polling state."""
+        return False
+
+    @asyncio.coroutine
+    def async_added_to_hass(self):
+        """Register callbacks."""
+        state = yield from async_get_last_state(self.hass, self.entity_id)
+        if state:
+            self._state = state.state == STATE_ON
+
+        @callback
+        def template_light_state_listener(entity, old_state, new_state):
+            """Handle target device state changes."""
+            # Determine need to evaluate template before returning.
+            yield_for_render = (self._template is not None or
+                                self._level_template is not None)
+            self.hass.async_add_job(
```

Comment:
I think I understand the confusion now. This can always be `async_update_ha_state(True)`, since if we don't change the internal attributes here, updating the ha state will never do anything.

But, you're right that we shouldn't ignore it either. What we should do is if self._template and self._level_template are both None, we shouldn't even schedule these updates [here](https://github.com/home-assistant/home-assistant/pull/7657/files#diff-267a90626dbaaf2cf7ffef1cc928a482R166). There's no reason to listen for any state changes if there are no templates that need to be re-rendered.